pkg/kubelet/rkt/rkt.go

```diff
@@ -0,0 +1,183 @@
+)
+
+const (
+	acversion            = "0.5.1"
+	kubernetesUnitPrefix = "k8s"
+	systemdServiceDir    = "/run/systemd/system"
+
+	rktDataDir             = "/var/lib/rkt"
+	rktLocalConfigDir      = "/etc/rkt"
+	rktMetadataServiceFile = "rkt-metadata.service"
+	rktMetadataSocketFile  = "rkt-metadata.socket"
+
+	unitKubernetesSection = "X-Kubernetes"
+	unitPodName           = "POD"
+	unitRktID             = "RktID"
+
+	dockerPrefix = "docker://"
+)
+
+const (
+	rktBinName = "rkt"
+
+	Embryo         = "embryo"
+	Preparing      = "preparing"
+	AbortedPrepare = "aborted prepare"
+	Prepared       = "prepared"
+	Running        = "running"
+	Deleting       = "deleting" // This covers pod.isExitedDeleting and pod.isDeleting.
+	Exited         = "exited"   // This covers pod.isExited and pod.isExitedGarbage.
+	Garbage        = "garbage"
+)
+
+const (
+	dockerAuthTemplate = `{"rktKind":"dockerAuth","rktVersion":"v1","registries":[%q],"credentials":{"user":%q,"password":%q}}`
+)
+
+var (
+	systemd *dbus.Conn
+	absPath string
+)
+
+// Runtime implements the ContainerRuntime for rkt. The implementation
+// uses systemd, so in order to run this runtime, systemd must be installed
+// on the machine.
+type Runtime struct {
+	systemd *dbus.Conn
+	absPath string
+	config  *Config
+	// TODO(yifan): Refactor this to be generic keyring.
+	dockerKeyring credentialprovider.DockerKeyring
+}
+
+func startMetadataService(systemd *dbus.Conn) error {
+	units, err := systemd.ListUnits()
+	if err != nil {
+		return err
+	}
+	for _, u := range units {
+		if u.Name == rktMetadataServiceFile {
+			// Metadata is already running.
+			return nil
+		}
+	}
+
+	// Create the service and socket file under systemd directory.
+	var src, dst string
+	dst = path.Join(systemdServiceDir, rktMetadataServiceFile)
+	if _, err := os.Stat(dst); err == nil {
+		if err := os.Remove(dst); err != nil {
+			return err
+		}
+	}
+	dst = path.Join(systemdServiceDir, rktMetadataSocketFile)
+	if _, err := os.Stat(dst); err == nil {
+		if err := os.Remove(dst); err != nil {
+			return err
+		}
+	}
+
+	wd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	dst = path.Join(systemdServiceDir, rktMetadataServiceFile)
+	src = path.Join(wd, rktMetadataServiceFile)
+	if err := os.Symlink(src, dst); err != nil {
+		return err
+	}
+	dst = path.Join(systemdServiceDir, rktMetadataSocketFile)
+	src = path.Join(wd, rktMetadataSocketFile)
+	if err := os.Symlink(src, dst); err != nil {
+		return err
+	}
+	_, err = systemd.StartUnit(rktMetadataServiceFile, "replace")
+	return err
+}
+
+// init will start the metadata-service as a transient systemd service.
+func init() {
+	var err error
+	systemd, err = dbus.New()
+	if err != nil {
+		glog.Errorf("rkt: Cannot connect to dbus: %v", err)
+	}
+
+	// Test if rkt binary is in $PATH.
+	absPath, err = exec.LookPath(rktBinName)
+	if err != nil {
+		glog.Errorf("rkt: Cannot find rkt binary: %v", err)
+	}
+
+	if err := startMetadataService(systemd); err != nil {
```

Comment:
We could really just send this in over go-systemd but we can optimise that later.